### PR TITLE
feat(#24): supports a common way for deployment waiting

### PR DESCRIPTION
**Pull request by anurnomeru**
#24

### CODE REVIEW

File: pkg/chunkserver/chunkserver.go [hunk 1]
```diff
@@ -106,14 +107,19 @@ func (c *Cluster) Start(nodeNameIP map[string]string) error {
...
 		return errors.Wrap(err, "failed to start chunkserver")
 	}
 
 	// 4. wait all chunkservers online before create logical pool
 	logger.Info("starting all chunkserver")
-	time.Sleep(30 * time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	if succeed := <-k8sutil.WaitForDeploymentsToStart(ctx, c.context.Clientset, 3*time.Second, chunkServers...); !succeed {
+		// TODO: should do anything here?
```
Comment (anurnomeru) [inline pkg/chunkserver/chunkserver.go:121]: should do anything here?

**Comment by anurnomeru** [comment]
I wrote a waiting case for one of the resources(chunkServers). If there is no problem, I will apply the changes the others.

File: pkg/chunkserver/chunkserver.go [hunk 2]
```diff
@@ -106,14 +107,19 @@ func (c *Cluster) Start(nodeNameIP map[string]string) error {
 	logger.Info("create physical pool successed")
 
 	// 3. startChunkServers start all chunkservers for each device of every node
-	err = c.startChunkServers()
+	var chunkServers []*v1.Deployment
```
Comment (caoxianfei1) [inline pkg/chunkserver/chunkserver.go:110]: I think this logic should not be added here. All processing is completed in startChunks, and the startChunkServers function should be that every deployment has started successfully. 

https://github.com/opencurve/curve-operator/blob/ebde98e0ef8d203c9eea8f78f44b2b78ab5ee14b/pkg/chunkserver/spec.go#L64

File: pkg/chunkserver/spec.go [hunk 3]
```diff
@@ -6,7 +6,8 @@ import (
 
 	"github.com/pkg/errors"
 	apps "k8s.io/api/apps/v1"
-	v1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
```
Comment (caoxianfei1) [inline pkg/chunkserver/spec.go:10]: v1 represent corve1 and apps/v1 is appsv1 may be better.

File: pkg/chunkserver/spec.go [hunk 4]
```diff
@@ -63,12 +65,11 @@ func (c *Cluster) startChunkServers() error {
 		} else {
 			logger.Infof("Deployment %s has been created , waiting for startup", newDeployment.GetName())
 			// TODO:wait for the new deployment
-			// deploymentsToWaitFor = append(deploymentsToWaitFor, newDeployment)
+			results = append(results, newDeployment)
 		}
 		// update condition type and phase etc.
 	}
-
```
Comment (caoxianfei1) [inline pkg/chunkserver/spec.go:70]: yes you can add your function here.

File: pkg/k8sutil/deployment.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+// WaitForDeploymentsToStart waits for the deployments to start, and returns a channel to indicate whether
+// all deployments are started or not
+//
+// tickDuration is the interval to check the deployment status
+// objectMeta is the metadata of the deployment
+//
+// we use the hub chan to collect the result of each deployment, and when all deployments are started,
+// we return true, otherwise, we return false, this design let WaitForDeploymentToStart and
+// WaitForDeploymentsToStart can be used in the same way
+func WaitForDeploymentsToStart(ctx context.Context, clientSet kubernetes.Interface, tickDuration time.Duration,
```
Comment (caoxianfei1) [inline pkg/k8sutil/deployment.go:21]: pass parameters using []*appsv1.Deployment, not ...

**Comment by caoxianfei1** [comment]
using go routine and channel is't must and you can handle it in main routine.
you can refer a function here, it more elegant.
https://github.com/rook/rook/blob/537c328c9722c4a77ab8ba96d7935be72ca250cd/pkg/operator/k8sutil/deployment.go#L118

**Comment by caoxianfei1** [comment]
@anurnomeru There is no need to use goroutines to detect. If you use coroutines, you must use channel, which will bring complexity and difficulty in understanding for others.

**Comment by caoxianfei1** [comment]
@anurnomeru you can continue it but i has add a simple implement for debug and test.

**Pull request closed without merging**